include/pybind11/detail/internals.h

```diff
@@ -150,7 +151,11 @@ struct type_info {
 };
 
 /// Tracks the `internals` and `type_info` ABI version independent of the main library version
+#ifndef PYBIND11_USE_SMART_HOLDER_AS_DEFAULT
 #define PYBIND11_INTERNALS_VERSION 4
+#else
+#define PYBIND11_INTERNALS_VERSION 1004
```

Comment:
Not sure if I 100% understand what "incompatible" means from PR overview.

Is it just that `pybind11` doesn't store `typeid` info for the holder type, so code using `py::class_<>` across diff. modalities (non-SH vs SH) may get messed up, so a new flavor of #1138?

If so, kewl for now! Just wanna confirm; perhaps worth adding a checkbox / issue / TODO next to this?
Ideally, the TODO would be removed when SH gets merged as a mainline option.

Reply:
I'm not 100% certain myself, but my understanding is that modules built with differing `PYBIND11_INTERNALS_VERSION` are fully isolated from each other. If that's true:

* _Without_ this PR, you'd indeed see a new flavor of #1138 (holder mismatch, leading to a `reinterpret_cast` with UB).
* But with this PR, types `py::class_`-wrapped in module_compiled_with_smart_holder_as_default are fully isolated from `py::class_` types wrapped in module_compiled_with_std_unique_ptr_as_default.

Does that sound plausible?

The upshot is that two modules that both use a certain `py::class_<T>`-wrapped type need to be built with matching `PYBIND11_USE_SMART_HOLDER_AS_DEFAULT`, or the casters for `T` will only be available in the module that has the `py::class_<T>`.

Reply:
Yup, totes sounds plausible! Would it be possible to encode this in a 1-sentence TODO or something?

Perhaps also mentioning obvious downshot: Because you're isolated, you can't share instances.

Reply:
Thanks for confirming! I'll rebase this PR and will add a small section to README_smart_holder.rst.

Reply:
@EricCousineau-TRI The "1-sentence TODO or something" became the longest section in README_smart_holder.rst :-)

**Classic / Conservative / Progressive cross-module compatibility**

I'll merge this PR now because we agree on the code change. I'll update the documentation as needed in separate PRs. As before, feedback for the documentation is highly welcome, chat, comment on this PR, new PR, whatever fits best.

Reply:
Sounds great, and sorry to motivate a large section haha! Thanks!